### PR TITLE
Hercules P32 mapping and Controls JS library

### PR DESCRIPTION
**Pull request by Be-ing**
This PR contains both a mapping for the Hercules P32 controller and a new JavaScript library to make mapping development easier. I'd like feedback on both. Documentation for the P32 functionality is on
[the wiki](http://mixxx.org/wiki/doku.php/hercules_p32_dj). I have started documenting the library inline with the code. Later, I'll move the library documentation to the wiki, but I think it should stay inline while this PR is being reviewed. Please ask questions about the code so I know what to document. :) Look at the P32-specific code for examples of how to use the library. Almost all functionality in the P32 mapping uses the library, except in some cases where it would be overcomplicated. If you have ideas of how to make the library more generally useful for different controllers, please comment.

The library is everything in Hercules-P32-scripts.js above `var P32 = {};`. For now, I think it should stay there. If anyone else wants to use it, feel free to copy and paste it into your own mapping. This will allow for rapid changes that don't have to maintain backwards compatibility. Perhaps after several people have used it for different controllers, it may make sense to move it to a separate shared JavaScript file.

The library depends on two functions from [Lodash](http://lodash.com/), [merge](https://lodash.com/docs/4.17.2#merge) and [assign](https://lodash.com/docs/4.17.2#assign). I made a custom build of Lodash including only those two functions. If you want to use the library yourself, you'll have to include the new lodash.mixxx.js file with your XML file.

The library is not compatible with Mixxx 2.0 because it takes advantage of functionality introduced by PR #919. I have made a Mixxx 2.0 compatible version of the P32 mapping that is functionally identical to this on [a separate branch](https://github.com/Be-ing/mixxx/tree/hercules_p32_mapping).

### CODE REVIEW

**Comment by Be-ing** [comment]
Documentation for each part of the library has been added inline. I'll continue to refine the language, but at this point, people should be able to read the documentation and understand how the library works and how to use it.

**Comment by Be-ing** [comment]
As a longer term goal, I'd like to move this towards a fully JS mapping system. The missing piece of that is being able to register MIDI input callbacks from JS rather than XML as discussed in PR #1016. The Control JS objects could be made to handle the callback objects in a similar manner as it currently handles the output callbacks.

**Comment by radusuciu** [comment]
I've noticed that in your XML you've script-bound everything. Is this strictly necessary? If I wanted to use your library, and only require a subset of the controls to be script-bound, is that something that will work well? I'd prefer to do this in the interest of taking advantage of the lower overhead of direct XML bindings where possible.

**Comment by Be-ing** [comment]
> I've noticed that in your XML you've script-bound everything. Is this strictly necessary?

To map it with a Control in the script, yes. You don't have to map everything that way, but I think it could be cumbersome to maintain if you mixed and matched which way parts of the controller were mapped.

> I'd prefer to do this in the interest of taking advantage of the lower overhead of direct XML bindings where possible.

I haven't noticed JS execution time for mappings to be an issue in any situation, even on my netbook with an Intel Atom N270 1.6 GHz CPU from 2008 (using much sloppier JS than this). Using this library generally requires less string manipulation per MIDI signal, as the string manipulation is generally done once when the Control object is created, then the I/O functions reference the stored properties of the object.

Regardless, the answer to performance concerns (if they actually matter) and the annoyance of duplicating everything in XML is to implement a way to register MIDI input callbacks from JS as noted above.

**Comment by Be-ing** [comment]
I have updated the effects mapping to make use of changes introduced by PR #1062 and #1063. A description is [on the wiki](http://mixxx.org/wiki/doku.php/hercules_p32_dj#mixxx_21). I have made it into a reusable ControlContainer for the library to make it easy to map controllers with similar effects sections.

**Comment by Be-ing** [comment]
[Wiki updated](http://mixxx.org/wiki/doku.php/hercules_p32_dj#effects) with description of new effects mapping.

**Comment by sblaisot** [comment]
Please add lodash.mixxx.js to windows uninstaller.

**Comment by Be-ing** [comment]
I think the EffectUnit ControlContainer is a good example of how this library can be used to manage the complex interaction of multiple layers in various conditions.

File: res/controllers/Hercules P32 DJ.midi.xml [hunk 1]
```diff
@@ -0,0 +1,2540 @@
+<?xml version='1.0' encoding='utf-8'?>
+<MixxxControllerPreset mixxxVersion="2.1.0+" schemaVersion="1">
+    <info>
+        <name>Hercules P32 DJ</name>
+        <author>Be</author>
+        <description>4-deck mapping for Hercules P32 controller</description>
+        <forums>http://mixxx.org/forums/viewtopic.php?f=7&amp;t=8132</forums>
```
Comment (rryan) [inline res/controllers/Hercules P32 DJ.midi.xml]: Want to add link to the wiki page for it?

**Comment by rryan** [comment]
thanks!

**Pull request merged**